src/Components/Components/src/Routing/InternalNavigationLock.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+/// <summary>
+/// A component that can be used to intercept internal navigation events.
+/// </summary>
+public class InternalNavigationLock : IComponent, IHandleLocationChanging, IDisposable
```

Comment:
In terms of API obviousness, could we consider naming this `NavigationLock` and having it handle both the internal and external cases? Maybe we wouldn't need the `NavigationPrompt` component to be separate.